Compiler/LangServer/BebopDiagnosticPublisher.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+using OmnisharpDiagnostic = OmniSharp.Extensions.LanguageServer.Protocol.Models.Diagnostic;
+
+namespace Compiler.LangServer
+{
+    public sealed class BebopDiagnosticPublisher
+    {
+        private readonly ILanguageServerFacade _facade;
+        private readonly BebopLangServerLogger _logger;
+
+        public BebopDiagnosticPublisher(
+            ILanguageServerFacade facade,
+            BebopLangServerLogger logger)
+        {
+            _facade = facade ?? throw new ArgumentNullException(nameof(facade));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public void ClearDiagnostics(DocumentUri uri, int? version)
+        {
+            _facade.TextDocument.PublishDiagnostics(new PublishDiagnosticsParams
+            {
+                Uri = uri,
+                Version = version,
+                Diagnostics = new Container<OmnisharpDiagnostic>(),
+            });
+        }
+
+        public void PublishDiagnostics(DocumentUri uri, int? version, IEnumerable<BebopDiagnostic> diagnostics)
+        {
+            // Group diagnostics by their filename
+            foreach (var group in diagnostics.GroupBy(x => x.Error.Span.FileName))
+            {
+                // If the schema is `unknown`, assume the diagnostics come from the currently opened file
+                var noSchema = group.Key == "(unknown)";
```

Comment:
Out of curiosity where does this string `"(unknown)"` come from?

Reply:
It seems to come from the schema reader: https://github.com/RainwayApp/bebop/blob/master/Core/IO/SchemaReader.cs#L34